### DISCUSSION TITLE
Support to skip client cert verification for redis on tls connection

### DISCUSSION THREAD

**Discussion by subhranil05** [juicedata/juicefs#3485] [General]
**What would you like to be added**:
We are using juciefs helm chart in our project where juicefs is connecting to a redis. Now we want to send tls connection by skiping client cert
**Why is this needed**:
Add a flag to skip client cert verification for redis on tls connection. We want to send tls connection by skiping client cert.
For other services there is a flag available to skip clinet cert verify.
We can add **tls-auth-clients** flag which is used in redis to skip client cert verification
I have tried with rediss://:****@redis-storage-xxxx:6379/0?tls-auth-clients=no
But it is saying as unsupported option
redis: unexpected option: tls-auth-clients [interface.go:409]

redis doc
https://redis.io/docs/management/security/encryption/

**Comment by yuhr123** [comment]
Redis is divided into TLS one-way authentication (only validates the server certificate) and mTLS mutual authentication (the server and client mutually verify each other's certificates) in terms of secure connections. Therefore, you need to confirm which authentication method is adopted by your Redis.

Suppose you use mTLS mutual authentication, you can disable the validation of client certificates by adding the `tls-auth-clients no` option in the redis.conf file.

**Comment by yuhr123** [comment]
By the way, if you want to pass your client certificates into your container, please refer to [how to set TLS for connect to Redis](https://juicefs.com/docs/community/databases_for_metadata/#set-up-tls)

**Comment by subhranil05** [comment]
@yuhr123 , thanks much for your response on this.
Yes my redis is using mTLS auth with `tls-auth-clients no` but I still need container certs, as I'm getting error for CA authority invalid cert when using only `rediss:// url` with no cert, 

Is there any way to pass insecure tls connection where I don't need to pass cert from container also and redis will not verify client cert as `tls-auth-clients no` ?

**Comment by subhranil05** [comment]
@zhijian-pro I have tried with  rediss://:****@redis-storage-xxxx:6379/0?insecure-skip-verify=true
not working. Erro details below
redis: unexpected option: insecure-skip-verify [interface.go:409]

**Comment by zhijian-pro** [comment]
> @zhijian-pro I have tried with rediss://:****@redis-storage-xxxx:6379/0?insecure-skip-verify=true not working. Erro details below redis: unexpected option: insecure-skip-verify [interface.go:409]

This is an unpublished feature of the main branch, and 'insecure-skip-verify=true' skips server certificate validation. Doesn't seem to solve your problem, but you can compile it and try.

**Comment by zhijian-pro** [comment]
@subhranil05  Can you provide the juicefs  output of error and  specific steps for reproducing the problem ？

**Comment by subhranil05** [comment]
@zhijian-pro I'm using juciefs helm chart where in command args section I'm using this
```
commands:
    - juicefs
args:
    - format
    - rediss://:****@redis-storage-xxxx:6379/0?insecure-skip-verify=true
```
but post install  hook` format-retain-volume` is failing as unsupported option with error ` redis: unexpected option: insecure-skip-verify [interface.go:409]` in pod's log

I have my `tls-auth-clients no` set in redis instance where juciefs is trying to connect
Please let me no if you need any additional info.

**Comment by subhranil05** [comment]
@zhijian-pro any luck?

**Comment by zhijian-pro** [comment]
@subhranil05  The `insecure-skip-verify=true` is an unpublished feature, which we will include in the 1.0.4 release tomorrow or the day after tomorrow, You can download 1.0.4 and then try `insecure-skip-verify=true`.
In addition, please provide the following error when metaurl is `rediss://:****@redis-storage-xxxx:6379/0` to help us analyze the problem.

**Comment by subhranil05** [comment]
@zhijian-pro thanks for your reply.
I'm using helm charts for `juicefs-csi-driver `and `juicefs-s3-gateway` .
Would you please help me to know the steps  in case of helm chart to get these changes?
Updating Docker images version to 1.0.4 and relevent in the chart should implement this change? Or need to update chart version?

**Comment by zhijian-pro** [comment]
@subhranil05  https://juicefs.com/docs/csi/guide/custom-image/    The latest tag is v1.0.4

**Comment by subhranil05** [comment]
@zhijian-pro We are using two charts `juicefs-csi-driver` and `juicefs-s3-gateway`

For `juicefs-csi-driver` we are using this connection block
```
commands:
    - juicefs
args:
    - format
    - rediss://:****@redis-storage-xxxx:6379/0
```
For `juicefs-s3-gateway` we are using this connection block
```
name: "juicefs-retain"
metaurl: rediss://:****@redis-storage-xxxx:6379/0
storage: "s3"
```
but it seems there is a post install hook for `juicefs-csi-driver` which is `format-retain-volume` is failing due to rediss and deployment is not going forward. So basically `juicefs-csi-driver`  is failing, please guide me to fix this
error details:
```
2023/04/07 08:30:50.177418 juicefs[1] <INFO>: Meta address: rediss://:****@redis-storage-xxxx:6379/0 [interface.go:402]
2023/04/07 08:30:55.266947 juicefs[1] <WARNING>: parse info: x509: certificate signed by unknown authority [redis.go:2868]
2023/04/07 08:30:55.272085 juicefs[1] <FATAL>: Load metadata: x509: certificate signed by unknown authority [format.go:411]
```

**Comment by zhijian-pro** [comment]
@subhranil05 
1. Try using juicefs 1.0.4 binary test directly to narrow down the problem first
2. `x509: certificate signed by unknown authority` may be because there is no 'ca-certificates' in the container. You can try to install them in the container first
